### ISSUE TITLE
Find CS of gt:condition interaction-mediating variants

### ISSUE THREAD

**Issue by JosephLalli** [stephenslab/mvsusieR#46]
Hi everyone!

I don't know whether this feature request is a whole new paper's worth of work, or a simple tweak to the existing algorithm. 

If I understand correctly, mash and mvSusie ask whether there is a CS that has an observed effect which is more likely to be drawn from the modeled distribution of effect variants than the modeled distribution of null variants. 

I would like determine whether a CS is more likely to be drawn from a modeled effect distribution than *the same CS in other, 'null' condition*. 

Another way to put it: I work on questions related to sex differences. I am less interested in whether or not a variant causes a change in phenotype, and I am more interested in whether a variant causes change in phenotype in one sex relative to another. 

If this were a simple linear regression, y ~ G + S + G:S, then susie finds the Bayesian equivalent of G, mvsusie finds the Bayesian equivalent of G + S, and I would like to find the Bayesian equivalent of G:S. 

If this is a substantial effort, then I will likely stick with my current approach (susie_rss using the interaction term's beta and se as bhat and shat). However, if this is a relatively small tweak to the current algorithm, I would greatly appreciate the feature.

Thanks,
Joe Lalli

**Comment by JosephLalli** [comment]
Would it be as simple as (in pseudocode):

fit$b1 = mean effect size
fit$b2 = variance of effect size
dimensions of each (since I forget what they actually are) = [#conditions, #variants, L]
for (set in fit$sets){
  L = set$L
  snps = CS$variants
  diff_beta = fit$b1[1, snps, L] - fit$b2[2, snps, L]
  diff_se = sqrt(fit$b2[1,snps, L]) + fit$b2[2, snps, L]
  diff_Z = diff_beta/diff_se
  diff_pval = pnorm(diff_Z, mean = 0, sd = 1, lower.tail = TRUE)
}

Or am I misunderstanding the situation, and a simple Z test won't apply in this setting? (Can you tell my stats are mostly self-taught?)

**Comment by gaow** [comment]
Hi Joe! It sounds like you might want to do stratified analysis by sex, get the CS globally via mvSuSiE, then use some metric to quantify if there is a condition specific effect? I tihnk mvSuSiE provides lfsr per condition that might be helpful. Otherwise, what we did in the MASH paper might be used here as well to test for sharing by sign or sharing by magnitude. I think this would be a good starting point rather than to worry about an interaction model? Of course there are more formal ways to test if the difference in the effect sizes are indeed significant across the two conditions but I think a good starting point is to quantify condition specific effect in mvSuSiE framework.

**Comment by gaow** [comment]
> Sharing by sign or by magnitude is closer, and the z test on the condition-specific-effect size (b1) and condition-specific effect variance (b2) is working ok, but I'm not sure if b1 and b2 are completely identical in concept to beta and variance.

We (a collaborative project that I lead) have been using something else along the same line of idea to quantify the difference between conditions. I'm very much preoccupied now but someone from my team is working on that. You can shoot me a message next week to touch base on it. We will eventually run it by the team here before suggesting the community to do that. Of course its just my two cents and perhaps @zouyuxin @pcarbo @stephens999 have other ideas

**Comment by stephens999** [comment]
I would think about this as a question about the difference in the effect size in males vs females. That is if b_j1 is effect in males of SNP j and b_j2 is effect in females, we can define the "clfsr" for the interaction term (analogous to equation (19) in https://www.biorxiv.org/content/10.1101/2023.04.14.536893v1)
as
 
clfsr_j  = 1- max(pr(b_j1-b_j2>0|...), pr(b_j1-b_j2<0|...))

If we computed these quantities then we could assign a measure of significance to each CS for the interaction term, in a similar way that we assign significance to each CS for each phenotype.

**Comment by JosephLalli** [comment]
> clfsr_j = 1- max(pr(b_j1-b_j2>0|...), pr(b_j1-b_j2<0|...))

This is precisely the quantity I would like to measure.

Can anyone send me a link to the line in mvSusie that calculates pr(b_jr > 0)? I don't know the code's organization particularly well, but I might be able to write a fork that would optionally calculate pr(b_jr-b_j1 > 0), where we assume the first effect is the baseline effect size.

**Comment by pcarbo** [comment]
I think Matthew meant equation 18 (which defines the clfsr), not equation 19 (which defines PIPs).

**Comment by pcarbo** [comment]
This sounds very similar to the setup for [MASH common baseline](https://stephenslab.github.io/mashr/articles/intro_mashcommonbaseline.html). Do I have this right?

(mvsusieR uses mash for the core computations, so presumably it should be possible to leverage mash common baseline as well in mvsusieR.)

**Comment by JosephLalli** [comment]
@pcarbo, I agree - that seemed like a perfect way to address my question! However, tl;dr - mashR's common baseline requires an R >= 3 to function.

Details: I've been working on this problem today, and I've noticed an issue: mashr, when adjusting for a common baseline condition using mash_update_data, reduces the number of conditions by one in the contrast_matrix function:

`if(ref %in% 1:R){
    L = diag(R)
    L[,ref] = -1
    L = L[-ref,]
    row.names(L) = paste0(name[-ref],'-', name[ref])`

and therefore in the mash_set_data_contrast function.

If L=2 in the above code, a 1D vector is returned.

`if(ref %in% 1:R){
        L = diag(R)
        L[,ref] = -1
        L = matrix(L[-ref,], nrow=R-1, ncol=R)
        row.names(L) = paste0(name[-ref],'-', name[ref])
    `

the result is still a Bhat/Shat with only one condition, which I don't believe mashR can work with.

What do you think is the best course of action? For now, I'm going to run mashR without a baseline and identify genes with male/female specific effects, then run mvSusie on those genes with a mashR-generated prior, and finally calculate a z-score on the resulting effect sizes.

This seems like a solvable problem, I just haven't spent as much time thinking about how to implement these ideas. @gaow @stephens999 , do you have any thoughts?

**Comment by stephens999** [comment]
definitely a solvable problem. Can you just run mash as usual and then run `mash_compute_posterior_matrices` with A = rbind(c(1,-1)) ?

**Comment by JosephLalli** [comment]
FWIW, I think I've figured out how to calculate the "likelihood R1 > or < R2" as a post-processing step on an mvSusie fit object. I recalculate mu and mu2 from b1/b2 and alpha, and then calculate the likelihood that the difference in means is positive using a z test. This doesn't really take correlation into account, but it will do for now.

A pseudocode example might look like this(my native coding language is python):
```
diff_likelihood <- function(fit, reference_group) {
        mu = fit$b1/fit$alpha
        mu2 = fit$b2/fit$alpha

        sigma = sqrt(mu2 - mu**2)

        mu_diff = mu[ , ,-reference_group] - mu[, , reference_group]
        sigma_diff = sqrt(sigma[, , -reference_group]^2 + sigma[,,reference_group]^2)

        pos_prob = pnorm(0, mean=mu_diff, sd=sigma_diff)
        pos_prob[is.na(pos_prob)] = 0
        neg_prob = 1-pos_prob
        
        pmax = max(pos_prob, neg_prob)
        sign_prob = sumrows((fit$alpha * pmax))
        lfsr_diff = clip(1 - sign_prob, 0, 1)
}
```
